api/v1alpha1/release_types_test.go

```diff
@@ -0,0 +1,446 @@
+			Expect(r.HasStarted()).To(BeFalse())
+		})
+
+		It("should return false when Status.startTime is zero", func() {
+			r.Status.StartTime = &metav1.Time{} // 0001-01-01 00:00:00 +0000 UTC
+			Expect(r.HasStarted()).To(BeFalse())
+		})
+
+		It("should return true when Status.startTime is not nil or zero", func() {
+			Expect(r.HasStarted()).To(BeTrue())
+		})
+	})
+
+	Context("When HasSucceeded method is called", func() {
+		It("should return false when condition.Status is ConditionUnknown", func() {
+			r.Status.Conditions[0] = metav1.Condition{
+				Type:   releaseConditionType,
+				Status: metav1.ConditionUnknown,
+			}
+			Expect(r.HasSucceeded()).To(BeFalse())
+		})
+
+		It("should return true when condition.Status is ConditionTrue", func() {
+			r.Status.Conditions[0] = metav1.Condition{
+				Type:   releaseConditionType,
+				Status: metav1.ConditionTrue,
+			}
+			Expect(r.HasSucceeded()).To(BeTrue())
+		})
+	})
+
+	Context("When IsDone method is called", func() {
+		It("should return false when condition struct is nil", func() {
+			Expect(r.IsDone()).To(BeFalse())
+		})
+
+		It("should return true when condition.Status is ConditionTrue", func() {
+			r.Status.Conditions[0] = metav1.Condition{
+				Type:   releaseConditionType,
+				Status: metav1.ConditionTrue,
+			}
+			Expect(r.IsDone()).To(BeTrue())
+		})
+
+		It("should return false when condition.Status is ConditionUnknown", func() {
+			r.Status.Conditions[0] = metav1.Condition{
+				Type:   releaseConditionType,
+				Status: metav1.ConditionUnknown,
+			}
+			Expect(r.IsDone()).To(BeFalse())
+		})
+	})
+
+	Context("When MarkFailed method is called", func() {
+		/*
+			|--------------------------------------------------------------------|
+			|                             NOOP matrix                            |
+			|--------------------------------------------------------------------|
+			| CompletionTime | condition.Status | time not nil | IsDone() | NOOP |
+			|----------------|------------------|--------------|----------|------|
+			|      zero      | ConditionUnknown |     true     |  false   |  no  |
+			|      zero      | ConditionTrue    |     true     |  true    |  yes |
+			|      nil       | ConditionUnknown |     false    |  false   |  no  |
+			|      nil       | ConditionTrue    |     false    |  true    |  no  |
+			|      Now()     | ConditionUnknown |     true     |  false   |  no  |
+			|      Now()     | ConditionTrue    |     true     |  true    |  yes |
+			|--------------------------------------------------------------------|
+
+			When no NOOP:
+			* expect zero or nil CompletionTime to be set to Now().
+			* expect setStatusConditionWithMessage() to append a new condition.
+			* * expect condition.Status: ConditionFalse
+			* * expect condition.Reason: to match the reason arg
+			* * expect condition.Message: to match the message arg
+		*/
+		It("should NOOP when IsDone() is true, and Status.CompletionTime is not nil", func() {
```

Comment:
We don't need that table on the top or any kind of comment. Let's rewrite this with something like:
```suggestion
		It("should do nothing if the Release has finished", func() {
```